stripe/src/test/java/com/stripe/android/StripeApiHandlerTest.java

```diff
@@ -265,6 +267,127 @@ public String getPackageName() {
+            PaymentIntentParams paymentIntentParams = PaymentIntentParams.createConfirmPaymentIntentWithSourceData(
+                    SourceParams.createCardParams(card),
+                    clientSecret,
+                    null
+            );
+            PaymentIntent paymentIntent = StripeApiHandler.confirmPaymentIntent(
+                    null,
+                    RuntimeEnvironment.application.getApplicationContext(),
+                    paymentIntentParams,
+                    publicKey,
+                    null,
+                    null);
+
+            assertNotNull(paymentIntent);
+
+        } catch (JSONException jsonEx) {
+            fail("Unexpected error: " + jsonEx.getLocalizedMessage());
+        } catch (AuthenticationException authEx) {
+            fail("Unexpected error: " + authEx.getLocalizedMessage());
+        } catch (StripeException stripeEx) {
+            fail("Unexpected error when connecting to Stripe API: "
+                    + stripeEx.getLocalizedMessage());
+        }
+    }
+
+    public void confirmPaymentIntent_withSourceId_canSuccessfulConfirm() {
+        String privateKey = "temporarily put a private key here simulate the backend";
+        String publicKey = "put a public key that matches the private key here";
+        try {
+            Card card = new Card("4242424242424242", 1, 2050, "123");
+            String jsonString = StripeApiHandler.createPaymentIntent(
+                    2500L,
+                    "usd",
+                    privateKey,
+                    null,
+                    SourceParams.createCardParams(card),
+                    null);
+
+            JSONObject jsonObject = new JSONObject(jsonString);
+
+            String clientSecret = jsonObject.optString("client_secret");
+            String sourceId = jsonObject.optString("source");
+            PaymentIntentParams paymentIntentParams = PaymentIntentParams.createConfirmPaymentIntentWithSourceId(
+                    sourceId,
+                    clientSecret,
+                    null
+            );
+            PaymentIntent paymentIntent = StripeApiHandler.confirmPaymentIntent(
+                    null,
+                    RuntimeEnvironment.application.getApplicationContext(),
+                    paymentIntentParams,
+                    publicKey,
+                    null,
+                    null);
+            assertNotNull(paymentIntent);
+
+        } catch (JSONException jsonEx) {
+            fail("Unexpected error: " + jsonEx.getLocalizedMessage());
+        } catch (AuthenticationException authEx) {
+            fail("Unexpected error: " + authEx.getLocalizedMessage());
+        } catch (StripeException stripeEx) {
+            fail("Unexpected error when connecting to Stripe API: "
+                    + stripeEx.getLocalizedMessage());
+        }
+    }
+
+    public void confirmRetrieve_withSourceId_canSuccessfulRetrieve() {
```

Comment:
You can test retrieve without creating first by providing the `client_secret` of a valid, canceled PaymentIntent. Feel free to re-use this publishable key + secret:

https://github.com/stripe/stripe-ios/blob/de5ce3c77314c4a303365b7e24f8cd6853189d2c/Tests/Tests/STPPaymentIntentFunctionalTest.m#L20-L25

stripe-js does the same thing, our understanding is that `client_secret` is valid for long periods of time.